Client/Frontend/Settings/CustomSearchEngineForm.swift

```diff
@@ -0,0 +1,188 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import UIKit
+import Shared
+import SnapKit
+import Storage
+import WebImage
+
+private let log = Logger.browserLogger
+
+class CustomSearchEngineForm: SettingsTableViewController {
+    
+    private var urlString: String?
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = "Add Custom Search Engine"
+    }
+    
+    func faviconFor(url: NSURL) -> Favicon? {
+        var resultFavicon: Favicon?
+        FaviconFetcher.getForURL(url, profile: profile).uponQueue(dispatch_get_main_queue()) { result in
+            guard let favicons = result.successValue where favicons.count > 0,
+                let _ = favicons.first?.url.asURL else {
+                    return
+            }
+            resultFavicon = favicons.first!
+        }
+        return resultFavicon
+    }
+    
+    func addSearchEngine(searchQuery: String) {
+        var iconURL: NSURL?
+        var iconImage: UIImage?
+        let SearchTermComponent = "%s"
+        let placeholder = "{searchTerms}"
+        var processedSearchQuery = ""
+        if searchQuery.componentsSeparatedByString(SearchTermComponent).count - 1 == 1{
+            processedSearchQuery = searchQuery.stringByReplacingOccurrencesOfString(SearchTermComponent, withString: placeholder)
+        }
+        guard processedSearchQuery != "",
+            let url = NSURL(string: processedSearchQuery.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.URLFragmentAllowedCharacterSet())!),
+            let shortName = url.domainURL().host
+            else {
+                let alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                self.presentViewController(alert, animated: true, completion: nil)
+                return
+        }
+        if let favicon = faviconFor(url) {
+            iconURL = NSURL(string: favicon.url)
+        }
+            iconImage = FaviconFetcher.getDefaultFavicon(url)
+        
+        let alert = ThirdPartySearchAlerts.addThirdPartySearchEngine { alert in
```

Comment:
An alert *is* only shown when adding the engine fails after tapping Save. This is how it was implemented in [https://github.com/mozilla-mobile/firefox-ios/blob/6a18631e369a4149796fff51b38b22e0b49b4477/Client/Frontend/Browser/BrowserViewController.swift#L3069](BrowserViewController).